tensorflow/python/layers/utils.py

```diff
@@ -178,61 +178,46 @@ def deconv_output_length(input_length, filter_size, padding, stride):
   return input_length
 
 
-def smart_cond(pred, fn1, fn2, name=None):
-  """Return either `fn1()` or `fn2()` based on the boolean predicate `pred`.
+def smart_cond(pred, true_fn=None, false_fn=None, name=None):
+  """Return either `true_fn()` if predicate `pred` is true else `false_fn()`.
 
-  If `pred` is a bool or has a constant value, we return either `fn1()`
-  or `fn2()`, otherwise we use `tf.cond` to dynamically route to both.
+  If `pred` is a bool or has a constant value, we return either `true_fn()`
+  or `false_fn()`, otherwise we use `tf.cond` to dynamically route to both.
 
   Arguments:
-    pred: A scalar determining whether to return the result of `fn1` or `fn2`.
-    fn1: The callable to be performed if pred is true.
-    fn2: The callable to be performed if pred is false.
+    pred: A scalar determining whether to return the result of `true_fn` or
+      `false_fn`.
+    true_fn: The callable to be performed if pred is true.
+    false_fn: The callable to be performed if pred is false.
     name: Optional name prefix when using `tf.cond`.
 
   Returns:
-    Tensors returned by the call to either `fn1` or `fn2`.
+    Tensors returned by the call to either `true_fn` or `false_fn`.
 
   Raises:
-    TypeError: If `fn1` or `fn2` is not callable.
+    TypeError: If `true_fn` or `false_fn` is not callable.
   """
-  if not callable(fn1):
-    raise TypeError('`fn1` must be callable.')
-  if not callable(fn2):
-    raise TypeError('`fn2` must be callable.')
-
-  pred_value = constant_value(pred)
-  if pred_value is not None:
-    if pred_value:
-      return fn1()
-    else:
-      return fn2()
-  else:
-    return control_flow_ops.cond(pred, fn1, fn2, name)
+  if isinstance(pred, variables.Variable):
+    return control_flow_ops.cond(pred, true_fn=true_fn, false_fn=false_fn, name=name)
+  return control_flow_ops.smart_cond(pred, true_fn=true_fn, false_fn=false_fn, name=name)
```

Comment:
Why not make smart_cond work with Variables too?

Reply:
I tried but I found “The problem is that variables already import control_flow_ops and control_flow_ops cannot import variables. I already resolved it without loop dependencies.”. This is commented in the former comments due to a test error. 
From the view the project structure, and also based on the original demands of boolean condition, I think it's ok to put it here.

Reply:
@sguada That is to say, `tensorflow/python/ops/variables.py` is using `cond` function of `tensorflow/python/ops/control_flow_ops.py`. If we do need consider using `Variables` class of `variables.py` in `control_flow_ops.py`. We may consider two solutions without circular dependencies:
1. Move `smart_cond` to some other places that's not depended by `tensorflow/python/ops/variables.py` and do that inclusion.
2. Just import that specific class `Variables` from `/tensorflow/python/ops.variables.py`

Otherwise, I think current version can at least work well with the demands of #13903. Let me know what you think is right and I will go for implementing that.